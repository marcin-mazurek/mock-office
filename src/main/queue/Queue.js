import unique from 'node-unique';
import Task from 'fun-task';
import { EventEmitter } from 'events';
import R from 'ramda';
import deepEqual from 'deep-equal';
import extractSubTree from './extractSubtree';

export const events = {
  TASK_REMOVED: 'TASK_REMOVED',
  TASK_RUN: 'TASK_RUN',
  TASK_ADDED: 'TASK_ADDED',
  TASK_REMOVED_AFTER_USE: 'TASK_REMOVED_AFTER_USE'
};

export default class Queue {
  constructor(args) {
    this.serverEE = args.ee;
    this.tasks = [];
    this.ee = new EventEmitter();
    this.runReadyTask = this.runReadyTask.bind(this);
    this.ee.on(events.TASK_REMOVED, this.runReadyTask);
    this.ee.on(events.TASK_RUN, this.runReadyTask);
    this.ee.on(events.TASK_ADDED, this.runReadyTask);
  }

  createTask(task) {
    let job;
    const t = task;

    if (t.interval) {
      job = Task.create(() => {
        t.running = true;
        const intervalId = setInterval(() =>
          this.tunnel({
            taskPayload: t.taskPayload,
            headers: t.headers
          }), t.interval
        );

        return () => {
          t.running = false;
          clearInterval(intervalId);
        };
      });
    } else if (t.delay) {
      job = Task.create((onSuccess) => {
        t.running = true;
        const timeoutId = setTimeout(() => {
          this.tunnel({
            taskPayload: t.taskPayload,
            headers: t.headers
          });
          onSuccess();
        }, t.delay);

        return () => {
          t.running = false;
          clearTimeout(timeoutId);
        };
      });
    } else {
      job = Task.create((onSuccess) => {
        t.running = true;
        this.tunnel({
          taskPayload: t.taskPayload,
          headers: t.headers
        });
        onSuccess();

        return () => {
          t.running = false;
        };
      });
    }

    return Object.assign(t, {
      id: unique(),
      job
    });
  }

  addTask(taskConfig) {
    const task = this.createTask(taskConfig);
    this.tasks.push(task);
    this.ee.emit(events.TASK_ADDED);

    return task.id;
  }

  runTask(taskId) {
    const task = this.tasks.find(exp => exp.id === taskId);

    task.cancel = task.job.run(() => {
      task.running = false;

      if (task.reuse === 'infinite') {
        R.identity(); // do nothing
      } else if (task.reuse === 'fixed') {
        task.quantity -= 1;

        if (task.quantity === 0) {
          this.removeTask(task.id);
          this.serverEE.emit(
            events.TASK_REMOVED_AFTER_USE,
            { queueId: this.id, taskId }
          );
        }
      } else {
        this.removeTask(task.id);
        this.serverEE.emit(events.TASK_REMOVED_AFTER_USE, { queueId: this.id, taskId });
      }
    });
  }

  removeTask(taskId) {
    const taskIndex = this.tasks.findIndex(task => task.id === taskId);
    const task = this.tasks[taskIndex];

    if (task.cancel) {
      task.cancel();
      task.cancel = undefined;
    }

    this.tasks.splice(taskIndex, 1);
    this.ee.emit(events.TASK_REMOVED);
  }

  findReadyTask(requirements) {
    for (let i = 0, len = this.tasks.length; i < len; i += 1) {
      const task = this.tasks[i];

      if (task.running) {
        if (task.blocking) {
          return undefined;
        }
      }

      if (task.requirements) {
        if (deepEqual(task.requirements, extractSubTree(requirements, task.requirements, {}))) {
          return task;
        }
      } else {
        return task;
      }
    }

    return undefined;
  }

  runReadyTask(requirements, cb) {
    if (this.tunnel) {
      const readyTask = this.findReadyTask(requirements);

      if (!readyTask) {
        if (cb) {
          cb();
        }
      } else {
        if (cb) {
          readyTask.job = readyTask.job.map(cb);
        }

        this.runTask(readyTask.id);
        this.ee.emit(events.TASK_RUN);
      }
    }
  }

  cancelPendingTasks() {
    this.tasks.forEach((task) => {
      if (task.cancel) {
        task.cancel();
      }
    });
  }

  openConnection(tunnel) {
    this.tunnel = tunnel;
  }

  closeConnection() {
    this.cancelPendingTasks();
    this.tunnel = undefined;
  }
}
