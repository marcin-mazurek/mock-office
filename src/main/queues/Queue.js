import unique from 'node-unique';
import Task from 'fun-task';
import { EventEmitter } from 'events';
import R from 'ramda';
import deepEqual from 'deep-equal';
import extractSubTree from './extractSubtree';
import { events as queuesEvents } from './index';

export const events = {
  TASK_REMOVED: 'TASK_REMOVED',
  TASK_RUN: 'TASK_RUN',
  TASK_ADDED: 'TASK_ADDED'
};

export default class Queue {
  constructor(args) {
    this.id = args.id;
    this.serverId = args.serverId;
    this.queuesEmitter = args.emitter;
    this.tasks = [];
    this.emitter = new EventEmitter();
    this.runReadyTasks = this.runReadyTasks.bind(this);
    this.emitter.on(events.TASK_REMOVED, this.runReadyTasks);
    this.emitter.on(events.TASK_RUN, this.runReadyTasks);
    this.emitter.on(events.TASK_ADDED, this.runReadyTasks);
  }

  createTask(task) {
    let job;
    const t = task;

    if (t.interval) {
      job = Task.create(() => {
        t.running = true;
        const intervalId = setInterval(() =>
          this.tunnel({
            taskPayload: t.taskPayload,
            headers: t.headers
          }), t.interval
        );

        return () => {
          t.running = false;
          clearInterval(intervalId);
        };
      });
    } else if (t.delay) {
      job = Task.create((onSuccess) => {
        t.running = true;
        const timeoutId = setTimeout(() => {
          this.tunnel({
            taskPayload: t.taskPayload,
            headers: t.headers
          });
          onSuccess();
        }, t.delay);

        return () => {
          t.running = false;
          clearTimeout(timeoutId);
        };
      });
    } else {
      job = Task.create((onSuccess) => {
        t.running = true;
        this.tunnel({
          taskPayload: t.taskPayload,
          headers: t.headers
        });
        onSuccess();

        return () => {
          t.running = false;
        };
      });
    }

    return Object.assign(t, {
      id: unique(),
      job
    });
  }

  addTask(taskConfig) {
    const task = this.createTask(taskConfig);
    this.tasks.push(task);
    this.emitter.emit(events.TASK_ADDED);

    return task.id;
  }

  runTask(taskId) {
    const task = this.tasks.find(exp => exp.id === taskId);

    task.cancel = task.job.run(() => {
      task.running = false;

      if (task.reuse === 'infinite') {
        R.identity(); // do nothing
      } else if (task.reuse === 'fixed') {
        task.quantity -= 1;

        if (task.quantity === 0) {
          this.removeTask(task.id);
          this.queuesEmitter.emit(
            queuesEvents.TASK_REMOVED_AFTER_USE,
            { queueId: this.id, taskId }
          );
        }
      } else {
        this.removeTask(task.id);
        this.queuesEmitter.emit(queuesEvents.TASK_REMOVED_AFTER_USE, { queueId: this.id, taskId });
      }
    });
  }

  removeTask(taskId) {
    const taskIndex = this.tasks.findIndex(task => task.id === taskId);
    const task = this.tasks[taskIndex];

    if (task.cancel) {
      task.cancel();
      task.cancel = undefined;
    }

    this.tasks.splice(taskIndex, 1);
    this.emitter.emit(events.TASK_REMOVED);
  }

  runReadyTasks(requirements, cb) {
    if (this.tunnel) {
      // collect phase
      let blockingTaskFound = false;
      const notBlockedTasks = R.takeWhile(
        (task) => {
          if (blockingTaskFound) {
            return false;
          }

          blockingTaskFound = task.behaviours;
          return true;
        },
        this.tasks
      );

      if (notBlockedTasks.length === 0) {
        if (cb) {
          cb();
        }

        return;
      }

      const runnableTasks = notBlockedTasks.filter((task) => {
        if (task.running) {
          return false;
        }

        // do not need
        if (!task.requirements) {
          return true;
        }

        // should fulfill requirements
        return deepEqual(task.requirements, extractSubTree(requirements, task.requirements, {}));
      });

      if (runnableTasks.length === 0) {
        if (cb) {
          cb();
        }

        return;
      }

      const firstRunnableTask = R.head(runnableTasks);

      if (cb) {
        firstRunnableTask.job.map(cb);
      }

      this.runTask(firstRunnableTask.id);
      this.emitter.emit(events.TASK_RUN);
    }
  }

  cancelPendingTasks() {
    this.tasks.forEach((task) => {
      if (task.cancel) {
        task.cancel();
      }
    });
  }

  openTunnel(tunnel) {
    this.tunnel = tunnel;
  }

  closeTunnel() {
    this.cancelPendingTasks();
    this.tunnel = undefined;
  }
}
