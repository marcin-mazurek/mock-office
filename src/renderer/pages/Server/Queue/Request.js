import React from 'react';

const HttpRequest = () => (
  <div>HttpRequest</div>
);

export default HttpRequest;
