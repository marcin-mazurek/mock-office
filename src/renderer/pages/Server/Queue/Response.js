import React from 'react';

const HttpResponse = () => (
  <div>HttpResponse</div>
);

export default HttpResponse;
