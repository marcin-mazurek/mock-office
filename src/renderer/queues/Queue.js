import React from 'react';
import TasksConnect from '../tasks/browseTasks/Tasks';

const Queue = ({ id }) => (
  <div className="queue">
    <div className="queue__tasks">
      <TasksConnect queueId={id} />
    </div>
  </div>
);

Queue.propTypes = {
  id: React.PropTypes.string.isRequired
};

export default Queue;
