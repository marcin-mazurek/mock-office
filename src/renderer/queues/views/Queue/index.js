import React from 'react';
import { connect } from 'react-redux';
import { getQueueResponses } from '../../selectors';
import Responses from '../../../responses/views/Responses';

const Queue = ({ responses, id }) => (
  <div className="queue">
    <div className="queue__responses">
      <Responses responsesIds={responses} queueId={id} />
    </div>
  </div>
);

Queue.propTypes = {
  responses: React.PropTypes.shape().isRequired,
  id: React.PropTypes.string.isRequired
};

const mapStateToProps = (state, ownProps) => ({
  responses: getQueueResponses(ownProps.id, state)
});

export default connect(mapStateToProps)(Queue);
