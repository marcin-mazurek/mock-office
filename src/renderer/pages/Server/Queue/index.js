import React from 'react';
import { connect } from 'react-redux';
import Request from './Request';
import Response from './Response';
import { getRequest, getResponses } from '../../../queues/selectors';

const HttpQueue = ({ request, responses }) => (
  <div>
    <div>
      <Request>{request.toString()}</Request>
    </div>
    <ul>
      {
        responses.map(response => (
          <Response response={response} />
        ))
      }
    </ul>
  </div>
);

const mapStateToProps = state => ({
  request: getRequest(state),
  responses: getResponses(state)
});

export default connect(mapStateToProps)(HttpQueue);
