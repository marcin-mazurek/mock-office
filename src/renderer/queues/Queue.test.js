import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Queue from './Queue';

test('Queue should render', () => {
  const wrapper = shallow(<Queue id="some-id" />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
