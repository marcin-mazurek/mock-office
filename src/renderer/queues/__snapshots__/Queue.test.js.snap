exports[`test Queue should render 1`] = `
<div
  className="queue">
  <div
    className="queue__tasks">
    <Connect(Component)
      serverId="some-id" />
  </div>
</div>
`;
